### ISSUE TITLE
INT8 Quantization Fails on Nvidia Jetson Xavier AGX (32GB) for ACT Model

### ISSUE THREAD

**Issue by ppangTae** [NVIDIA/TensorRT#4527]
## Description

Hello,

I attempted to perform INT8 quantization with TensorRT on an Action Chunking with Transformer (ACT) neural network model using Nvidia Jetson Xavier AGX 32GB, but the engine build fails with the following error:

[E] 2: [weightConvertors.cpp::quantizeBiasCommon::337] Error Code 2: Internal Error (Assertion getter(i) != 0 failed. )
[E] 2: [builder.cpp::buildSerializedNetwork::751] Error Code 2: Internal Error (Assertion engine != nullptr failed. )
[!] Invalid Engine. Please ensure the engine was built correctly
polygraphy.exception.exception.PolygraphyException: Invalid Engine. Please ensure the engine was built correctly

Interestingly, the same model builds successfully with INT8 quantization on my desktop environment with the following specs:(The only change I made was lowering the ONNX opset version from 20 to 16 during model export.)

    GPU: RTX 5070

    CUDA: 12.8

    TensorRT: 10.12

Inference works fine and performance improves as expected.

However, on the Jetson Xavier AGX, the engine build only works with FP32 — INT8 quantization fails with the error above.

## Environment
**TensorRT Version**: 8.5.2.2

**NVIDIA GPU**: Nvidia Jeston Xavier AGX 32GB

**NVIDIA Driver Version**: Jetpack 5.1.3

**CUDA Version**: 11.4
 
**CUDNN Version**: 8.6.0

python version :  3.8

## Relevant Files
https://drive.google.com/drive/folders/1GLMG8U0mfSbi0izat7xL6AXo3ouRBh7k?usp=sharing
act/act.onnx : onnx model (opset 16)
act/act.cache : calibrator cache file
act/act_origin.engine : fp32 engine

## onnx transformation code
```python
    batch_size = 1
    state_dim = 6
    num_cameras = 3
    channels = len(camera_names)
    height = 480
    width = 640

    dummy_qpos = torch.randn(batch_size, state_dim).cuda()
    dummy_image = torch.randn(batch_size, num_cameras, channels, height, width).cuda()

    # ONNX로 export
    onnx_path = os.path.join(save_dir, "act.onnx")
    torch.onnx.export(
        policy,                                
        (dummy_qpos, dummy_image),             
        onnx_path,                            
        export_params=True,                    
        opset_version=16,              
        do_constant_folding=True,           
        input_names=['qpos', 'image'],         
        output_names=['action'],                
    )
```

## Data Loader for INT8 Calibrator(polygraphy)
```python
def data_generator(norm_stats, representative_episode_ids):
    for episode_id in representative_episode_ids:
        dataset_path = os.path.join(dataset_dir, f"episode_{episode_id}.hdf5")
        with h5py.File(dataset_path, 'r') as root:
            qpos_data = root['/observations/qpos'][()] # shape : (episode_len, qpos)
            image_dict = dict()
            for cam_name in camera_names:
                image_dict[cam_name] = root[f'/observations/images/{cam_name}'][()] # shape : (episode_len, 480, 640, 3)

            # new axis for different cameras
            all_cam_images = []
            for cam_name in camera_names: 
                all_cam_images.append(image_dict[cam_name])
            all_cam_images = np.stack(all_cam_images, axis=0) # shape : (# of camera, episode_len, 480, 640 ,3)

            # channel last
            image_data = torch.from_numpy(all_cam_images)
            qpos_data = torch.from_numpy(qpos_data)
            image_data = torch.einsum('b k h w c -> k b c h w', image_data) # shape : (episode_len, # of camera, 3, 480, 640)

            # Change tensor to numpy
            image_data = image_data.numpy()
            qpos_data = qpos_data.numpy()

            # normalize image and change dtype to float
            image_data = image_data / np.float32(255.0)
            qpos_data = (qpos_data - norm_stats["qpos_mean"]) / norm_stats["qpos_std"]

        for qpos, image in zip(qpos_data, image_data):
            # Dict key must be the same as ONNX input name
            qpos = np.expand_dims(qpos, axis=0)
            image = np.expand_dims(image, axis=0)
            yield {"qpos" : qpos, "image" : image}
```

```python
    builder, network, parser = poly_trt.network_from_onnx_path(path=onnx_model_path)

    if fp32:
        builder_config = poly_trt.create_config(builder=builder,
                                                network=network)
    else:
        if calibration:
            calibrator = poly_trt.Calibrator(data_loader=data_generator(norm_stats, episode_ids),
                                             cache="act.cache")
            
        else:
            calibrator = poly_trt.Calibrator(data_loader=random_data_generator(norm_stats))
           

        # Each type flag must be set to true.
        builder_config = poly_trt.create_config(builder=builder, network=network,
                                                int8=True, fp16=fp16, tf32=tf32,
                                                calibrator=calibrator)

    engine = poly_trt.engine_from_network(network=(builder, network, parser), 
                                          config=builder_config)
```

Any insights into why this might be happening or how to resolve it would be greatly appreciated.

Thanks in advance!

**Comment by kevinch-nv** [comment]
Is TensorRT Version: 8.5.2.2 the correct version you're using on the Xavier board? 8.5 is quite old and is no longer supported. Can you update through the latest Jetpack version? https://developer.nvidia.com/embedded/jetpack

**Comment by krish-05-2002** [comment]
@ppangTae  — Were you able to successfully quantize the ACT model? If so, could you share any reference documentation or scripts you used for calibration?

@kevinch-nv  — I am working on INT8 calibration for both LLM and ACT models. Are there any official documents available for INT8 quantization of these models?

**Comment by kevinch-nv** [comment]
For LLMs specifically checkout https://github.com/NVIDIA/TensorRT-LLM.
For general quantization use https://github.com/NVIDIA/TensorRT-Model-Optimizer

**Comment by krish-05-2002** [comment]
@kevinch-nv 
@ppangTae 
# Title
ACT (Action Chunking with Transformers) INT8 engine has very low accuracy vs FP32 on RTX 3090 (Polygraphy / TensorRT calibration)

## Summary
I attempted to build an INT8 TensorRT engine (using Polygraphy) from an ONNX ACT (Action Chunking with Transformers) model. The engine builds successfully, but INT8 inference accuracy is significantly worse than FP32 results on an NVIDIA GeForce RTX 3090. The same ONNX exported model with FP32 gives expected results.

Even after retraining the model with more epochs, INT8 accuracy degradation persists, so it seems related to calibration/quantization rather than underfitting.

Below is a minimal repro script I used for calibration + engine creation and details of my environment. I would appreciate suggestions on how to increase INT8 accuracy.

---

## Repro script (calibration + build)
```
import os
import h5py
import torch
import numpy as np
from polygraphy.backend import trt as poly_trt
from polygraphy.backend.trt import CreateConfig, Profile, engine_from_network, network_from_onnx_path, save_engine, Calibrator

onnx_model_path = "act_model_tested_5000.onnx"   
engine_path = "act_model_int8_5000_test_3.engine"

dataset_dir = "act/tested_dataset"
camera_names = ["top"]

representative_episode_ids = list(range(50)) 

norm_stats = {
    "qpos_mean": np.array([
        -0.00537649, -0.4888646, 1.013746, -0.00614289, -0.5179883, 1.1218338,
        0.68953073, 0.01946618, -0.31947428, 0.47818068, -0.02486075, 0.7694499,
        0.03743789, 0.70888525
    ]),
    "qpos_std": np.array([
        0.00364376, 0.5279966, 0.19803946, 0.01640265, 0.3591604, 0.59978694,
        0.2553522, 0.10964487, 0.5030749, 0.44584075, 0.14309423, 0.30574822,
        0.22783644, 0.24754341
    ]),
}

def data_generator(norm_stats, representative_episode_ids):
    for episode_id in representative_episode_ids:
        dataset_path = os.path.join(dataset_dir, f"episode_{episode_id}.hdf5")
        with h5py.File(dataset_path, 'r') as root:
            qpos_data = root['/observations/qpos'][()]  # (episode_len, qpos_dim)

            image_dict = {}
            for cam_name in camera_names:
                if f'/observations/images/{cam_name}' not in root:
                    raise KeyError(f"Camera {cam_name} not found in dataset!")
                image_dict[cam_name] = root[f'/observations/images/{cam_name}'][()]  

            all_cam_images = [image_dict[cam] for cam in camera_names]
            all_cam_images = np.stack(all_cam_images, axis=0)  

            image_data = torch.from_numpy(all_cam_images)  # (#cams, episode_len, H, W, 3)
            qpos_data = torch.from_numpy(qpos_data)

            image_data = torch.einsum('b k h w c -> k b c h w', image_data)

            image_data = image_data.numpy()
            qpos_data = qpos_data.numpy()

            image_data = (image_data / 255.0).astype(np.float32)
            qpos_data = (qpos_data - norm_stats["qpos_mean"]) / norm_stats["qpos_std"]
            qpos_data = qpos_data.astype(np.float32)

        for qpos, image in zip(qpos_data, image_data):
            qpos = np.expand_dims(qpos, axis=0)     # (1, qpos_dim)
            image = np.expand_dims(image, axis=0)   # (1, #cams, 3, H, W)

            yield {"qpos": qpos, "image": image}
def build_int8_engine():
    # Load ONNX model into TRT network
    builder, network, parser = network_from_onnx_path(onnx_model_path)

    # Create calibrator
    calibrator = Calibrator(
        data_loader=data_generator(norm_stats, representative_episode_ids),
        cache="act_model_test_3.cache"
    )

    # Enable INT8 + (optionally FP16/TF32)
    builder_config = CreateConfig(
        int8=True,
        fp16=False,   # set True if you want mixed-precision FP16+INT8
        tf32=False,
        calibrator=calibrator
    )

    # Build engine
    engine = engine_from_network(
        (builder, network, parser),
        config=builder_config
    )

    # Save engine
    save_engine(engine, engine_path)
    print(f"[INFO] INT8 engine saved to {engine_path}")

if __name__ == "__main__":
    build_int8_engine()
```

# Observed Behavior

- FP32 & FP16 engine → good accuracy (matches training expectations).
- INT8 engine → large accuracy drop.

# Tried So Far

- Increased number of calibration samples (50 episodes).
- Retrained model with more epochs.
- Used cache for calibration (.cache file).

# Question
What are the recommended best practices for improving INT8 accuracy in TensorRT when calibrating vision+robotics models like ACT?

- Should I increase dataset size
- Would enabling mixed precision (INT8 + FP16) help mitigate loss?
